src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -166,315 +190,344 @@ public void emit(Event event)
-        throw new IOException(String.format("Timed out after [%d] millis during flushing", config.getFlushTimeOut()), e);
-      }
+  private void flush(Batch batch) throws IOException
+  {
+    if (batch == null) {
+      return;
+    }
+    batch.seal();
+    try {
+      emittedBatchCounter.awaitBatchEmitted(batch.batchNumber, config.getFlushTimeOut(), TimeUnit.MILLISECONDS);
+    }
+    catch (TimeoutException e) {
+      String message = String.format("Timed out after [%d] millis during flushing", config.getFlushTimeOut());
+      throw new IOException(message, e);
+    }
+    catch (InterruptedException e) {
+      log.debug("Thread Interrupted");
+      Thread.currentThread().interrupt();
+      throw new IOException("Thread Interrupted while flushing", e);
     }
   }
 
   @Override
   @LifecycleStop
   public void close() throws IOException
   {
-    synchronized (started) {
-      // flush() doesn't do things if it is not started, so flush must happen before we mark it as not started.
-      flush();
-      started.set(false);
-      exec.shutdown();
+    synchronized (startLock) {
+      if (started) {
+        started = false;
+        Batch lastBatch = concurrentBatch.getAndSet(null);
+        flush(lastBatch);
+        // Emitter is interrupted after the last batch is flushed.
+        emittingThread.interrupt();
+      }
     }
   }
 
-  private class EmittingRunnable implements Runnable
+  private class EmittingThread extends Thread
   {
-    private final long instantiatedVersion;
+    private ZeroCopyByteArrayOutputStream gzipBaos;
 
-    public EmittingRunnable(
-        long instantiatedVersion
-    )
+    EmittingThread()
     {
-      this.instantiatedVersion = instantiatedVersion;
+      super("HttpPostEmitter-" + instanceCounter.incrementAndGet());
+      setDaemon(true);
     }
 
     @Override
     public void run()
     {
-      long currVersion = version.get();
-
-      try {
-        if (!started.get()) {
-          log.info("Not started, skipping...");
-          return;
+      while (true) {
+        boolean needsToShutdown = needsToShutdown();
+        try {
+          emitLargeEvents();
+          emitBatches();
         }
-
-        if (instantiatedVersion != currVersion) {
-          log.debug("Skipping because instantiatedVersion[%s] != currVersion[%s]", instantiatedVersion, currVersion);
+        catch (Throwable t) {
+          log.error(t, "Uncaught exception in EmittingThread.run()");
+        }
+        if (needsToShutdown) {
+          // Make GC life easier
+          drainBuffersToReuse();
           return;
-        } else {
-          count.set(0);
-          currVersion = version.incrementAndGet();
         }
+        // Waiting for 1/2 of config.getFlushMillis() in order to flush events not more than 50% later than specified.
+        // If nanos=0 parkNanos() doesn't wait at all, that we don't want.
+        long waitNanos = Math.max(TimeUnit.MILLISECONDS.toNanos(config.getFlushMillis()) / 2, 1);
```

Comment:
Finish sleeping here doesn't mean automatic emit, it additionally checks conditions in `needsToShutdown()` method, where it doesn't half the FlushMillis period, but emits anyway, if the emitterThread is interrupted, that is done in `close()`, i. e. it means the service is in the process of shutdown. So events are not regularly emitted sooner than FlushMillis, unless the service is shutting down, when emitting sooner is a good thing.

I actually added this check after some tests were failing, waiting for the last events (after httpPostEmitter.close() is called) longer than expected.